### DISCUSSION TITLE
Would it be possible for the blog plugin to `setGlobalData({blogPosts})`?

### DISCUSSION THREAD

**Discussion by ilg-ul** [facebook/docusaurus#9922] [Q&A]
With the current release code, I could not find a way to access a selection of the blog posts in standalone or docs pages.

Would it be possible for the blog plugin to save the blog posts globally (with `setGlobalData()`) and make the content available outside the blog plugin? Similar question for tags, authors, and possibly other data.

I saw a notice in the documentation (_Global data is... global: its size affects the loading time of all pages of your site, so try to keep it small._), but I don't understand how the loading time of a static site can be affected by some JSON files saved during build time.

If saving the blogPosts in the blog plugin is not possible, is it reasonable to do this in the new `processBlogPosts()` hook? What solutions would be possible for the other data (tags, authors, etc)?

**Comment by Josh-Cena** [comment]
> but I don't understand how the loading time of a static site can be affected by some JSON files saved during build time.

The global data is not just "saved as a JSON file". It is loaded globally in your site via `const globalData = require("global-data.json")` or something similar, so that anywhere in your site, you can use `useGlobalData` to retrieve this data. This JSON data directly lives in your main client bundle.

We most definitely do not want to save literally everything about blog posts globally. Nevertheless I think `processBlogPosts` could be a solution.

**Reply by ilg-ul** [reply]
> We most definitely do not want to save literally everything about blog posts globally.

Why?

**Reply by Josh-Cena** [reply]
Because that means everything about your blog posts are loaded globally, even those that you don't use, and those that most people will never use.

**Comment by slorber** [comment]
> I don't understand how the loading time of a static site can be affected by some JSON files saved during build time.

This is because you don't understand yet what React hydration is. Learn that and it will make sense.

---

Global data need to be loaded on every single page of your site before React could hydrate, because hydration assumes the data is loaded, because we use that data to display things such as the global site layout features (version dropdowns, i18n dropdowns etc).

If this data is very heavy, this delays the React hydration. Users on a bad network will encounter a bad UX where the site/page is visible, but it is not possible to interact with it for a few seconds. If users click on buttons, and the buttons do nothing, this is a really bad UX. Increasing global data size increase the time a site is presented as non-interactive, and should be avoided.

Now if you don't mind, you can extend the blog plugin `contentLoaded` to add the global data yourself. Just keep in mind that all this data needs to be loaded before the page becomes interactive, so it depends on the size you plan to put there.

https://github.com/facebook/docusaurus/issues/4138


```tsx
const blogPluginExports = require("docusaurus-plugin-content-blog");

const defaultBlogPlugin = blogPluginExports.default;

async function blogPluginExtended(...pluginArgs) {
  const blogPluginInstance = await defaultBlogPlugin(...pluginArgs);
  const pluginOptions = pluginArgs[1];
  return {
    ...blogPluginInstance,
    contentLoaded: async function (params) {

      // Register yourself the extra global data you want here
      const { content, actions } = params;
      actions.setGlobalData(content);
      
      return blogPluginInstance.contentLoaded(params);
    },
  };
}

module.exports = {
  ...blogPluginExports,
  default: blogPluginExtended,
};
```

---

Another possibility is to have the blog plugin create additional pages, I've explained this setup here:
https://github.com/facebook/docusaurus/issues/4138#issuecomment-1703083513

This is more efficient (no global data usage) but will not work in all situations (particularly if you need data from 2 distinct plugins on the same page).

---

I doubt `processBlogPosts` would solve this problem and wonder what makes you think it could.

---

The solution we want to this "problem" (being able to inject any plugin data to anywhere) is React Server Components running at build time. But it's complicated to implement and we haven't started working on this yet.

https://github.com/facebook/docusaurus/issues/9089

Until we have this, I'll try to figure out if we can provide a temporary good enough workaround.

**Reply by ilg-ul** [reply]
> This is because you don't understand yet what React hydration is. Learn that and it will make sense.

I could not find any useful references to React hydration in the Docusaurus documentation. :-(

I think you should at least mention it in the documentation site, explain why Docusaurus needs it, and provide a link to an external site where it is documented properly, possibly with examples how this is used by Docusaurus, such that Docusaurus users can understand it and make use of it.

Searching for 'react hydration' on Google was not very useful, it returned something about Gatsby, a deprecated [page](https://react.dev/reference/react-dom/hydrate#), and a bunch of other unintelligible pages.

Probably anyone knows what React hydration is and I'm not smart enough to get it... :-(

**Reply by slorber** [reply]
React hydration is not a Docusaurus concept, and regular Docusaurus users (those using core plugins and writing markdown files) do not really need to be aware of what it is. Now if you want to understand better how Docusaurus actually works under the hood, knowing what React hydration is very important, and you can find many resources online by searching "What is Reaact hydration" on google.

Docusaurus is very similar to Gatsby, and their docs page explains it:
https://www.gatsbyjs.com/docs/conceptual/react-hydration/

(note: there's a reason for Gatsby to explain it, and Docusaurus not: Gatsby is a more technical non-opinionated framework where all users have to be actual React devs, unlike Docusaurus)


---

I agree that the React docs site doesn't really explain this concept, but that's not our fault 😅 cc @mattcarrollcode that's probably worth having a canonical resource explaining what hydration is overall? And not just an `hydrateRoot` API page?

Until then, some resources you could read:
- https://react.dev/reference/react-dom/client/hydrateRoot
- https://blog.saeloun.com/2021/12/16/hydration/
- https://www.joshwcomeau.com/react/the-perils-of-rehydration/
- https://blog.logrocket.com/react-hydration-pre-rendered-html/
- https://www.builder.io/blog/hydration-is-pure-overhead

**Reply by ilg-ul** [reply]
Ok, so (what I hoped to be) a static site is actually full of .js code, and saving global data basically increases the size of this code, which increases the time to read it in, which makes the site less responsive. BTW, 'rehydration' is an unfortunate name.

**Reply by slorber** [reply]
Not all static sites, but this applies to React-based static/SSR sites indeed.

https://blog.saeloun.com/2021/12/16/hydration/

![image](https://github.com/facebook/docusaurus/assets/749374/211ed908-9dfb-41e6-a05b-d6ebc66d2a92)

Other static sites might use Vanilla JS or another solution instead of React. Loading that JS may not be called hydration, but the same problem is here, and the site is not "interactive" until the required JS loads.

**Reply by ilg-ul** [reply]
In the end I don't mind Docusaurus using this technique in order to provide interactivity.

But I think that not providing a convenient way to share data between plugins **at build time** is a Docusaurus limitation.

I don't know, perhaps all my use cases are weird, being originated in how Jekyll works, with all data stored in global arrays.

My first attempt to use Docusaurus was to migrate about 20+ small project documentation sites from Jejyll to Docusaurus. It failed because in one of the standalone pages I need a simple list of releases, which are blog posts with a `release` tag (there are other posts too, with various project announcements).

These pages look like this:

![Screenshot 2024-03-07 at 12 39 04](https://github.com/facebook/docusaurus/assets/362732/8342f18a-accb-450d-8506-4dd47a67f468)

I did not know how to do it with the public Docusaurus release, and the migration was postponed. But the problem did not go away, I still need to migrate those 20+ sites, and if I don't find a convenient solution, I might have to use a local fork of Docusaurus. Please note that swizzling the same components in an identical mode for 20+ projects is not exactly a joy.

The second attempt was the historical site, where, in the home page, I have a table of events (actually three tables, each on a separate tab):

![Screenshot 2024-03-07 at 12 44 20](https://github.com/facebook/docusaurus/assets/362732/9f2ba360-92a2-41d3-8ac5-3ef0b6f95e2c)

Same problem, the content comes from the blog posts. Again, I did not know how to do it with the public Docusaurus release, but this time I forked Docusaurus and added lots of hacks, just to make it work.

Given your latest explanations, the solution seems to be to extend the blog plugin `contentLoaded()` to prepare the data required in these tables (trying to keep these tables to a minimum to not impact responsiveness due to hydration) and add it to the global data. 

Can you confirm that this is the current recommended solution?

---

Notes:

- As discussed in another thread, a more convenient solution would be an `afterContentLoaded()` hook.
- My understanding of React is still very poor, so sorry if this is just another stupid question, but why isn't it possible to have something like `setBuildGlobalData()/useBuildGlobalData()` that happens only at build time, and is not propagated into the generated site?

**Reply by slorber** [reply]
> But I think that not providing a convenient way to share data between plugins at build time is a Docusaurus limitation.

We all agree on that, it's a known problem. Docusaurus plugins are kind-of sandboxed from one another.

> I don't know, perhaps all my use cases are weird, being originated in how Jekyll works, with all data stored in global arrays.

Docusaurus is not Jekyll, because it's a SPA giving you a more stateful app and soft navigation instead of full page reloads:
- Jekyll only has to render HTML files at build time, and do not need to send this data to the frontend
- Docusaurus has to render HTML both at build time, and inside the browser when you click links, so the browser should download everything it needs to render the next page, in an efficient way. This usually gives a better UX, but is also more challenging to implement technically.



Gatsby used to add GraphQL queries alongside React components, letting you retrieve the data you need from a larger "graphql data lake". This looked like a good idea in 2015 when it came out, but it turns out it quickly became complicated and now devs do not like Gatsby because of that query layer.

```tsx
import * as React from 'react'
import { graphql } from 'gatsby'

const HomePage = ({data}) => {
  return (
    <div>
     {data.site.siteMetadata.description}
    </div>
  )
}

export const query = graphql`
  query HomePageQuery {
    site {
      siteMetadata {
        description
      }
    }
  }
`

export default HomePage
```

Gatsby inspired the creation of Docusaurus v2, but the complex GraphQL layer that the community didn't like was dropped. Docusaurus is simpler but does not currently have a good data layer, to let you inject data from one plugin to another.

React Server Components are going to be the best solution to this problem, letting you "pick" the right data you need directly inside React components, using TS/JS object accesses instead of a query system.


---

I'm thinking of a possible good enough temporary solution.

The `contentLoaded` actually takes a 3rd undocumented arg `allContent`: https://docusaurus.io/docs/api/plugin-methods/lifecycle-apis#contentLoaded

```
async contentLoaded({actions: {createData, addRoute}, allContent})
```

Our debug plugin is currently able to create pages by reading content from other plugins:
https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-debug/src/index.ts


This is undocumented because it's still considered internal, and I don't like this API, and the fact we have to "wait" for all plugin data to finally be able to call `contentLoaded` is not great either for optimizing execution times.

And this also prevents me from enabling even faster fine-grained hot reloads I'm working on in https://github.com/facebook/docusaurus/pull/9903
(not reloading whole site on change, but just a single affected plugin)

For these reasons I'm planning to introduce a new `allContentLoaded` lifecycle method asap. That would enable you to read from all the site data and create pages with it, or generating hardcoded json files that you could then read in React components directly through `import myJson from @generated/myJsonFile`

**Reply by ilg-ul** [reply]
> For these reasons I'm planning to introduce a new allContentLoaded lifecycle method asap.

Great! Any ETE?

**Reply by slorber** [reply]
That's blocking me from completing https://github.com/facebook/docusaurus/pull/9903 properly (and it's the thing I'm working on atm) so very likely tomorrow 😄

**Reply by slorber** [reply]
Not sure it will perfectly resolve your problem (it's not the main motivation of this new lifecycle) but here's the PR: https://github.com/facebook/docusaurus/pull/9931

**Reply by ilg-ul** [reply]
Let me see if I understood it. For the historical site, I have to define a custom blog plugin, and implement the `allContentLoaded()` function where I get all blogPosts, I process them to compute the content for the chronology tables, I use `createData()` and `addRoute()` (as you explained in https://github.com/facebook/docusaurus/issues/4138#issuecomment-1703083513) to generate the home page?

Has the size of these JSON files generated by `createData()` any impact on site responsiveness? In other words, is it better to do all formatting before creating the data, or is it acceptable to pass some raw properties (that take some more space) and do the formatting in the component that displays the tables?

If I manage to do this, I guess that for the 20+ small sites it'll be easier to apply the same recipe (filter the blog posts, get those tagged with 'release' and use `createData()` and `addRoute()` to generate that specific page).

BTW, I did read the explanations in https://github.com/facebook/docusaurus/pull/9931 several times, and it is not clear for me why it was necessary to add `allContentLoaded()`, and not pass the same data as `allContent` via `contentLoaded()`.

Also the text mentions that this new feature allows to aggregate data from multiple plugins, but it does not provide an example how to do it. (Currently I don't need this, the content of my chronology tables comes from a single blog plugin, but in my site I have 3 blog instances, and I might want to show data from the other blogs too).

**Comment by slorber** [comment]
> Let me see if I understood it. For the historical site, I have to define a custom blog plugin, and implement the `allContentLoaded()` function where I get all blogPosts, I process them to compute the content for the chronology tables, I use `createData()` and `addRoute()` (as you explained in [#4138 (comment)](https://github.com/facebook/docusaurus/issues/4138#issuecomment-1703083513)) to generate the home page?



You do not even have to extend the blog plugin to add your own `allContentLoaded`, since all plugins receive all the data, this can as well be a pure standalone plugin.




> Has the size of these JSON files generated by `createData()` any impact on site responsiveness? In other words, is it better to do all formatting before creating the data, or is it acceptable to pass some raw properties (that take some more space) and do the formatting in the component that displays the tables?


It has impact only on the pages you inject that data into. If your homepage receives a lot of data, your homepage will be heave. Unlike global data this size will not impact ALL your pages, but just this one.


> BTW, I did read the explanations in #9931 several times, and it is not clear for me why it was necessary to add `allContentLoaded()`, and not pass the same data as `allContent` via `contentLoaded()`.

That is related to the DX optimizations I'm working on as part of https://github.com/facebook/docusaurus/pull/9903. If `contentLoaded` received that data, then on any file edit, we'd have to run this lifecycle for ALL the plugins. What we want instead is that if you edit a blog post, we only have to call the affected blog instance `contentLoaded()` method, and we do not have to call the docs `contentLoaded()`. This method creates files and can be expensive so we should try to only call it when needed. Most plugins do not need to read content from one another. If you do not understand this part it's not a problem but it will enable me to give a much faster hot reload experience on file edits, that would otherwise be unsafe.


> Also the text mentions that this new feature allows to aggregate data from multiple plugins, but it does not provide an example how to do it. (Currently I don't need this, the content of my chronology tables comes from a single blog plugin, but in my site I have 3 blog instances, and I might want to show data from the other blogs too).

Agree, here's a full example that I'll also add to the PR:

https://stackblitz.com/edit/github-agngt8?file=MyHomePage%2FMyHomePagePlugin.js,MyHomePage%2FMyHomePage.js,docusaurus.config.js

![CleanShot 2024-03-14 at 16 28 14@2x](https://github.com/facebook/docusaurus/assets/749374/538302b5-8997-44f8-9d65-09242b96bc15)

```js
export default async function MyHomePagePlugin(context, options) {
  return {
    name: 'my-home-page-plugin',
    async allContentLoaded({ allContent, actions }) {
      const allBlogs = allContent['docusaurus-plugin-content-blog'];
      const blog1 = allBlogs['blog1'];
      const blog2 = allBlogs['blog2'];
      const allBlogPosts = [...blog1.blogPosts, ...blog2.blogPosts];

      actions.addRoute({
        path: '/',
        exact: true,
        component: '@site/MyHomePage/MyHomePage.js',
        modules: {
          allBlogPosts: await actions.createData(
            'allBlogPosts.json', // The file name doesn't matter, we'll simplify this API soon
            JSON.stringify(allBlogPosts)
          ),
        },
      });
    },
  };
}
```

```js
export default function Home(props) {
  const { allBlogPosts } = props;
  return (
      <main>
        <h2>All blog posts</h2>
        <ul>
          {allBlogPosts.map((blogPost, i) => (
            <li key={i}>{blogPost.metadata.title}</li>
          ))}
        </ul>
      </main>
  );
}
```

**Reply by ilg-ul** [reply]
That was helpful, thank you!
